File: Typescript/Ejercicio 3/flujo_juego.ts
import { GestionPilaCartas } from "./gestion_pila_cartas";
import { InterfazConsola } from "./interfaz_consola";
import * as fs from "fs";

interface Jugador {
    nombre: string;
    puntos: number;
    activo: boolean;
}

class FlujoJuego {
    private jugadores: Jugador[] = [];
    private pila: GestionPilaCartas;
    private interfaz: InterfazConsola;

    constructor() {
        this.pila = new GestionPilaCartas();
        this.interfaz = new InterfazConsola();
    }

    iniciarJuego() {
        const n = this.interfaz.pedirNumeroJugadores();
        for (let i = 0; i < n; i++) {
            const nombre = this.interfaz.pedirNombreJugador(i);
            this.jugadores.push({ nombre, puntos: 0, activo: true });
        }
        this.ronda(this.jugadores);
    }

    private ronda(participantes: Jugador[]) {
        this.interfaz.mostrarMensaje("--- Nueva Ronda ---");
        participantes.forEach(jugador => {
            jugador.puntos = 0;
            jugador.activo = true;

            while (jugador.activo) {
                const carta = this.pila.sacarCarta();
                if (!carta) break;
                let valor = this.valorCarta(carta, jugador.puntos);
                jugador.puntos += valor;

                this.interfaz.mostrarMensaje(
                    `${jugador.nombre} obtuvo ${carta} (total: ${jugador.puntos})`
                );

                if (jugador.puntos > 21) {
                    this.interfaz.mostrarMensaje(`${jugador.nombre} se pasó de 21 y queda eliminado`);
                    jugador.activo = false;
                    break;
                }

                if (!this.interfaz.preguntarCarta(jugador.nombre)) {
                    jugador.activo = false;
                }
            }
        });

        // Evaluar ganador o empate
        const vivos = participantes.filter(j => j.puntos <= 21);
        if (vivos.length === 0) {
            this.interfaz.mostrarMensaje("Nadie gana en esta ronda.");
            return;
        }

        const maxPuntos = Math.max(...vivos.map(j => j.puntos));
        const empatados = vivos.filter(j => j.puntos === maxPuntos);

        if (empatados.length === 1) {
            const ganador = empatados[0];
            this.interfaz.mostrarMensaje(`¡${ganador.nombre} gana con ${ganador.puntos} puntos!`);
            fs.writeFileSync("ganadores.txt", `Ganador: ${ganador.nombre} (${ganador.puntos})\n`, { flag: "a" });
        } else {
            this.interfaz.mostrarMensaje(`Empate entre: ${empatados.map(j => j.nombre).join(", ")}`);
            this.ronda(empatados); 
        }
    }

    private valorCarta(carta: string, puntosActuales: number): number {
        if (["J", "Q", "K"].includes(carta)) return 10;
        if (carta === "A") return puntosActuales + 11 <= 21 ? 11 : 1;
        return parseInt(carta);
    }
}

// Ejecutar el juego automáticamente si se llama desde Node
if (require.main === module) {
    const juego = new FlujoJuego();
    juego.iniciarJuego();
}

File: Go/Ejercicio 3/flujo_juego.go
package main

import (
	"fmt"
	"os"
)

type Jugador struct {
	Nombre string
	Puntos int
	Activo bool
}

type FlujoJuego struct {
	jugadores []Jugador
	pila      *GestionPilaCartas
	interfaz  *InterfazConsola
}

func NuevoFlujoJuego() *FlujoJuego {
	return &FlujoJuego{
		pila:     NuevaGestionPilaCartas(),
		interfaz: &InterfazConsola{},
	}
}

func (f *FlujoJuego) IniciarJuego() {
	n := f.interfaz.PedirNumeroJugadores()
	for i := 0; i < n; i++ {
		nombre := f.interfaz.PedirNombreJugador(i)
		f.jugadores = append(f.jugadores, Jugador{Nombre: nombre, Puntos: 0, Activo: true})
	}
	f.Ronda(f.jugadores)
}

func (f *FlujoJuego) Ronda(participantes []Jugador) {
	f.interfaz.MostrarMensaje("--- Nueva Ronda ---")

	for i := range participantes {
		participantes[i].Puntos = 0
		participantes[i].Activo = true

		for participantes[i].Activo {
			carta, ok := f.pila.SacarCarta()
			if !ok {
				break
			}
			valor := f.valorCarta(carta, participantes[i].Puntos)
			participantes[i].Puntos += valor
			f.interfaz.MostrarMensaje(fmt.Sprintf("%s obtuvo %s (total: %d)", participantes[i].Nombre, carta, participantes[i].Puntos))

			if participantes[i].Puntos > 21 {
				f.interfaz.MostrarMensaje(fmt.Sprintf("%s se pasó de 21 y queda eliminado", participantes[i].Nombre))
				participantes[i].Activo = false
				break
			}

			if !f.interfaz.PreguntarCarta(participantes[i].Nombre) {
				participantes[i].Activo = false
			}
		}
	}

	var vivos []Jugador
	for _, j := range participantes {
		if j.Puntos <= 21 {
			vivos = append(vivos, j)
		}
	}

	if len(vivos) == 0 {
		f.interfaz.MostrarMensaje("Todos perdieron")
		return
	}

	maxPuntos := vivos[0].Puntos
	for _, j := range vivos {
		if j.Puntos > maxPuntos {
			maxPuntos = j.Puntos
		}
	}

	var empatados []Jugador
	for _, j := range vivos {
		if j.Puntos == maxPuntos {
			empatados = append(empatados, j)
		}
	}

	if len(empatados) == 1 {
		ganador := empatados[0]
		f.interfaz.MostrarMensaje(fmt.Sprintf("¡%s gana con %d puntos!", ganador.Nombre, ganador.Puntos))

		file, _ := os.OpenFile("ganadores.txt", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
		defer file.Close()
		fmt.Fprintf(file, "Ganador: %s (%d)\n", ganador.Nombre, ganador.Puntos)
	} else {
		nombres := ""
		for _, j := range empatados {
			nombres += j.Nombre + " "
		}
		f.interfaz.MostrarMensaje("Empate entre: " + nombres)
		f.Ronda(empatados)
	}
}

func (f *FlujoJuego) valorCarta(carta string, puntosActuales int) int {
	switch carta {
	case "J", "Q", "K":
		return 10
	case "A":
		if puntosActuales+11 <= 21 {
			return 11
		}
		return 1
	default:
		var v int
		fmt.Sscan(carta, &v)
		return v
	}
}

func main() {
	juego := NuevoFlujoJuego()
	juego.IniciarJuego()
}
